pkg/controllers/edgeconnect/controller_test.go

```diff
@@ -219,6 +223,87 @@ func TestReconcile(t *testing.T) {
+		ec := &edgeconnect.EdgeConnect{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      testName,
+				Namespace: testNamespace,
+			},
+			Spec: edgeconnect.EdgeConnectSpec{
+				ApiServer: "abc12345.dynatrace.com",
+				OAuth: edgeconnect.OAuthSpec{
+					Endpoint:     "https://test.com/sso/oauth2/token",
+					Resource:     "urn:dtenvironment:test12345",
+					ClientSecret: testOauthClientSecret,
+					Provisioner:  false,
+				},
+			},
+		}
+		controller := createFakeClientAndReconciler(t, ec,
+			createClientSecret(testOauthClientSecret, ec.Namespace),
+			createKubeSystemNamespace(),
+		)
+
+		_, err := controller.Reconcile(context.TODO(), reconcile.Request{
+			NamespacedName: types.NamespacedName{Namespace: testNamespace, Name: testName},
+		})
+		require.NoError(t, err)
+
+		err = controller.apiReader.Get(context.TODO(), client.ObjectKey{Name: ec.Name, Namespace: ec.Namespace}, ec)
+		require.NoError(t, err)
+		require.NotEmpty(t, ec.Conditions())
+
+		condition := meta.FindStatusCondition(*ec.Conditions(), consts.SecretConfigConditionType)
+		assert.Equal(t, metav1.ConditionTrue, condition.Status)
+		assert.Equal(t, conditions.SecretCreatedReason, condition.Reason)
+		assert.Equal(t, ec.Name+"-"+consts.EdgeConnectSecretSuffix+" created", condition.Message)
+	})
+
+	t.Run(`SecretConfigConditionType is set SecretGenFailed`, func(t *testing.T) {
+		ec := &edgeconnect.EdgeConnect{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      testName,
+				Namespace: testNamespace,
+			},
+			Spec: edgeconnect.EdgeConnectSpec{
+				ApiServer: "abc12345.dynatrace.com",
+				OAuth: edgeconnect.OAuthSpec{
+					Endpoint:     "https://test.com/sso/oauth2/token",
+					Resource:     "urn:dtenvironment:test12345",
+					ClientSecret: testOauthClientSecret,
+					Provisioner:  false,
+				},
+			},
+		}
+
+		controller := createFakeClientAndReconciler(t, ec,
+			createClientSecret(testOauthClientSecret, ec.Namespace),
+			createKubeSystemNamespace(),
+		)
+		fakeClient := fake.NewClientWithInterceptors(interceptor.Funcs{
+			Get: func(ctx context.Context, client client.WithWatch, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+				if _, ok := obj.(*corev1.Secret); ok {
+					return fmt.Errorf("get secret error")
+				}
+
+				return nil
+			},
+		})
+		controller.apiReader = fakeClient
+
+		err := controller.reconcileEdgeConnectRegular(context.TODO(), ec)
+		require.Error(t, err)
+
+		err = controller.apiReader.Get(context.TODO(), client.ObjectKey{Name: ec.Name, Namespace: ec.Namespace}, ec)
+		require.NoError(t, err)
+		require.NotEmpty(t, ec.Conditions())
+
+		condition := meta.FindStatusCondition(*ec.Conditions(), consts.SecretConfigConditionType)
+		assert.Equal(t, metav1.ConditionFalse, condition.Status)
+		assert.Equal(t, conditions.SecretGenerationFailed, condition.Reason)
+		assert.Equal(t, "Failed to generate secret: get secret error", condition.Message)
+	})
```

Comment:
Isn't this what you want?
```suggestion
	t.Run("SecretConfigConditionType is set SecretGenFailed", func(t *testing.T) {
		ec := &edgeconnect.EdgeConnect{
			ObjectMeta: metav1.ObjectMeta{
				Name:      testName,
				Namespace: testNamespace,
			},
			Spec: edgeconnect.EdgeConnectSpec{
				ApiServer: "abc12345.dynatrace.com",
				OAuth: edgeconnect.OAuthSpec{
					Endpoint:     "https://test.com/sso/oauth2/token",
					Resource:     "urn:dtenvironment:test12345",
					ClientSecret: testOauthClientSecret,
					Provisioner:  false,
				},
			},
		}

		controller := createFakeClientAndReconciler(t, ec,
			createKubeSystemNamespace(),
		)

		err := controller.reconcileEdgeConnectRegular(context.Background(), ec)
		require.Error(t, err)
		require.NotEmpty(t, ec.Conditions())

		condition := meta.FindStatusCondition(*ec.Conditions(), consts.SecretConfigConditionType)
		assert.Equal(t, metav1.ConditionFalse, condition.Status)
		assert.Equal(t, conditions.SecretGenerationFailed, condition.Reason)
		assert.Contains(t, condition.Message, "Failed to generate secret: failed to get clientSecret from client-secret secret")
	})
```

Reading the test is super confusing:
1. Here you configure a `controller` with an `apiReader` that will have the client-secret
```
		controller := createFakeClientAndReconciler(t, ec,
			createClientSecret(testOauthClientSecret, ec.Namespace),
			createKubeSystemNamespace(),
		)
```
2. Here you override the `apiReader`, to always die in case someone wants to get an error, and `return nil` other wise which is incorrect
```
		fakeClient := fake.NewClientWithInterceptors(interceptor.Funcs{...})
		controller.apiReader = fakeClient
```
- You made the previous call`createFakeClientAndReconciler` basically useless, as the whole purpose of the func is to setup the `apiReader` mock
- When overwriting a `Get` with an interceptor, you should make sure that all other cases still work (ie.: that is realistic)
   - Your `Get` interceptor will do impossible things when getting an edgeconnect resource for example (you do this in the next step)
     - What is impossible? : Doing nothing(ie.: not touching the `obj client.Object`) and returning `nil` error as a result of a `Get`. This will never-ever happen when using the actual `client`
3. Here you get the edgeconnect from the API, even though the function you are testing will not update the edgeconnect CR in the API, but only locally
```
		err = controller.apiReader.Get(context.TODO(), client.ObjectKey{Name: ec.Name, Namespace: ec.Namespace}, ec)
		require.NoError(t, err)
		require.NotEmpty(t, ec.Conditions())
```
- But why does the test pass?
  - Because of the weird `interceptor.Funcs` you defined for the `Get` action will do nothing, therefore the `ec` object here is actually left untouched, so the local changes done in the `reconcileEdgeConnectRegular` func will be present

Reply:
What my suggestion basically does:
1. In the `createFakeClientAndReconciler` call, I leave out the config-secret, so it wont be present, so the logic will fail, where you want it to fail
2. Removed the `interceptor.Funcs` part, as it is unnecessary
3. Removed the `get` call for the `ec` as the function we are testing modifies the object, and doesn't update it on the API
4. Not checking the exact error message as that is annoying (used `assert.Contains`)
5. Used `context.Background` instead of TODO

Reply:
I am pretty sure you could simplify the other tests for this function, as it reads very much like copy paste